v0.18.3.4 FTBFS: Arch Linux

```
[dormouse@psappho monero]$ git clean -fxd
Removing build/
[dormouse@psappho monero]$ mkdir build
[dormouse@psappho monero]$ cd build/
[dormouse@psappho build]$ cmake -Wno-dev .. > cmake.log 2>&1
[dormouse@psappho build]$ make -j3 > make.log 2>&1
[dormouse@psappho build]$ git status
HEAD detached at v0.18.3.4
nothing to commit, working tree clean
```
cmake.log:
https://plch.xyz/static/base_app/cmake.log

make.log:
https://plch.xyz/static/base_app/make.log

This looks like uuid related error. 

I was under impression that we merged a fix [1]. 

This was due to (breaking) update to `boost::uuid` in latest boost. I assume you are using a recent version of boost. 

Let me double check it. 

```
home/dormouse/gits/monero/contrib/epee/include/span.h:165:34: error: static assertion failed: source type may have padding
  165 |     static_assert(!has_padding<T>(), "source type may have padding");
      |                    ~~~~~~~~~~~~~~^~
/home/dormouse/gits/monero/contrib/epee/include/span.h:165:34: note: â! epee::has_padding<boost::uuids::uuid>()â evaluates to false
make[2]: *** [src/rpc/CMakeFiles/obj_rpc.dir/build.make:107: src/rpc/CMakeFiles/obj_rpc.dir/core_rpc_server.cpp.o] Error 1
```


1. https://github.com/monero-project/monero/pull/9462

Don't close yet, I have another failure, I pressed `comment` way too soon.

```
[ 51%] Built target obj_p2p
[ 51%] Building CXX object src/rpc/CMakeFiles/obj_rpc.dir/instanciations.cpp.o
In file included from /home/dormouse/gits/monero/src/cryptonote_core/tx_pool.h:47,
                 from /home/dormouse/gits/monero/src/cryptonote_core/cryptonote_core.h:45,
                 from /home/dormouse/gits/monero/src/cryptonote_protocol/cryptonote_protocol_handler-base.cpp:52:
/home/dormouse/gits/monero/src/cryptonote_basic/cryptonote_basic_impl.h:43:29: warning: ‘template<class _Arg, class _Result> struct std::unary_function’ is deprecated [-Wdeprecated-declarations]
   43 |   struct array_hasher: std::unary_function<t_array&, std::size_t>
      |                             ^~~~~~~~~~~~~~
In file included from /usr/include/c++/14.2.1/string:49,
                 from /home/dormouse/gits/monero/src/cryptonote_protocol/cryptonote_protocol_handler-base.cpp:33:
/usr/include/c++/14.2.1/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
      |            ^~~~~~~~~~~~~~
[ 51%] Building CXX object src/cryptonote_protocol/CMakeFiles/obj_cryptonote_protocol.dir/levin_notify.cpp.o
/home/dormouse/gits/monero/src/cryptonote_protocol/block_queue.cpp:45:21: error: redefinition of ‘struct std::hash<boost::uuids::uuid>’
   45 |   template<> struct hash<boost::uuids::uuid> {
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/boost/uuid/nil_generator.hpp:8,
                 from /home/dormouse/gits/monero/src/cryptonote_protocol/block_queue.cpp:33:
/usr/include/boost/uuid/uuid.hpp:353:19: note: previous definition of ‘struct std::hash<boost::uuids::uuid>’
  353 | template<> struct hash<boost::uuids::uuid>
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~
make[3]: *** [src/cryptonote_protocol/CMakeFiles/obj_cryptonote_protocol.dir/build.make:79: src/cryptonote_protocol/CMakeFiles/obj_cryptonote_protocol.dir/block_queue.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
[ 51%] Building CXX object src/rpc/CMakeFiles/obj_rpc_pub.dir/zmq_pub.cpp.o
In file included from /home/dormouse/gits/monero/src/cryptonote_basic/cryptonote_format_utils.h:33,
                 from /home/dormouse/gits/monero/src/rpc/zmq_pub.cpp:48:
/home/dormouse/gits/monero/src/cryptonote_basic/cryptonote_basic_impl.h:43:29: warning: ‘template<class _Arg, class _Result> struct std::unary_function’ is deprecated [-Wdeprecated-declarations]
   43 |   struct array_hasher: std::unary_function<t_array&, std::size_t>
      |                             ^~~~~~~~~~~~~~
In file included from /usr/include/c++/14.2.1/string:49,
                 from /usr/include/boost/assert/source_location.hpp:13,
                 from /usr/include/boost/exception/exception.hpp:9,
                 from /usr/include/boost/throw_exception.hpp:21,
                 from /usr/include/boost/thread/pthread/mutex.hpp:12,
                 from /usr/include/boost/thread/mutex.hpp:16,
                 from /home/dormouse/gits/monero/src/rpc/zmq_pub.h:33,
                 from /home/dormouse/gits/monero/src/rpc/zmq_pub.cpp:30:
/usr/include/c++/14.2.1/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
      |            ^~~~~~~~~~~~~~
[ 51%] Built target obj_rpc_pub
[ 51%] Building CXX object src/serialization/CMakeFiles/obj_serialization.dir/json_object.cpp.o
In file included from /home/dormouse/gits/monero/src/serialization/json_object.cpp:36:
/home/dormouse/gits/monero/src/cryptonote_basic/cryptonote_basic_impl.h:43:29: warning: ‘template<class _Arg, class _Result> struct std::unary_function’ is deprecated [-Wdeprecated-declarations]
   43 |   struct array_hasher: std::unary_function<t_array&, std::size_t>
      |                             ^~~~~~~~~~~~~~
In file included from /usr/include/c++/14.2.1/string:49,
                 from /usr/include/boost/utility/string_ref_fwd.hpp:19,
                 from /usr/include/boost/utility/string_ref.hpp:22,
                 from /home/dormouse/gits/monero/src/serialization/json_object.h:31,
                 from /home/dormouse/gits/monero/src/serialization/json_object.cpp:29:
/usr/include/c++/14.2.1/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
      |            ^~~~~~~~~~~~~~
[ 51%] Built target obj_serialization
[ 51%] Building CXX object src/rpc/CMakeFiles/obj_daemon_messages.dir/message.cpp.o
[ 51%] Building CXX object src/rpc/CMakeFiles/obj_daemon_messages.dir/daemon_messages.cpp.o
In file included from /home/dormouse/gits/monero/src/cryptonote_core/tx_pool.h:47,
                 from /home/dormouse/gits/monero/src/cryptonote_core/cryptonote_core.h:45,
                 from /home/dormouse/gits/monero/src/rpc/core_rpc_server.h:42,
                 from /home/dormouse/gits/monero/src/rpc/core_rpc_server.cpp:38:
/home/dormouse/gits/monero/src/cryptonote_basic/cryptonote_basic_impl.h:43:29: warning: ‘template<class _Arg, class _Result> struct std::unary_function’ is deprecated [-Wdeprecated-declarations]
   43 |   struct array_hasher: std::unary_function<t_array&, std::size_t>
      |                             ^~~~~~~~~~~~~~
In file included from /usr/include/c++/14.2.1/string:49,
                 from /usr/include/c++/14.2.1/bits/locale_classes.h:40,
                 from /usr/include/c++/14.2.1/locale:41,
                 from /usr/include/boost/filesystem/path.hpp:22,
                 from /usr/include/boost/filesystem.hpp:16,
                 from /home/dormouse/gits/monero/src/rpc/core_rpc_server.cpp:33:
/usr/include/c++/14.2.1/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
      |            ^~~~~~~~~~~~~~
In file included from /home/dormouse/gits/monero/contrib/epee/include/hex.h:38,
                 from /home/dormouse/gits/monero/contrib/epee/include/string_tools.h:30,
                 from /home/dormouse/gits/monero/src/rpc/core_rpc_server.cpp:35:
/home/dormouse/gits/monero/contrib/epee/include/span.h: In instantiation of ‘epee::span<const unsigned char> epee::as_byte_span(const T&) [with T = boost::uuids::uuid]’:
/home/dormouse/gits/monero/contrib/epee/include/string_tools.h:94:39:   required from ‘std::string epee::string_tools::pod_to_hex(const t_pod_type&) [with t_pod_type = boost::uuids::uuid; std::string = std::__cxx11::basic_string<char>]’
   94 |     return to_hex::string(as_byte_span(s));
      |                           ~~~~~~~~~~~~^~~
/home/dormouse/gits/monero/src/rpc/core_rpc_server.cpp:3270:76:   required from here
 3270 |       const std::string span_connection_id = epee::string_tools::pod_to_hex(span.connection_id);
      |                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/span.h:165:34: error: static assertion failed: source type may have padding
  165 |     static_assert(!has_padding<T>(), "source type may have padding");
      |                    ~~~~~~~~~~~~~~^~
/home/dormouse/gits/monero/contrib/epee/include/span.h:165:34: note: ‘! epee::has_padding<boost::uuids::uuid>()’ evaluates to false
make[3]: *** [src/rpc/CMakeFiles/obj_rpc.dir/build.make:107: src/rpc/CMakeFiles/obj_rpc.dir/core_rpc_server.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
[ 52%] Building CXX object src/rpc/CMakeFiles/obj_daemon_rpc_server.dir/daemon_handler.cpp.o
In file included from /home/dormouse/gits/monero/src/cryptonote_core/tx_pool.h:47,
                 from /home/dormouse/gits/monero/src/cryptonote_core/cryptonote_core.h:45,
                 from /home/dormouse/gits/monero/src/p2p/net_node.inl:60,
                 from /home/dormouse/gits/monero/src/rpc/instanciations.cpp:32:
/home/dormouse/gits/monero/src/cryptonote_basic/cryptonote_basic_impl.h:43:29: warning: ‘template<class _Arg, class _Result> struct std::unary_function’ is deprecated [-Wdeprecated-declarations]
   43 |   struct array_hasher: std::unary_function<t_array&, std::size_t>
      |                             ^~~~~~~~~~~~~~
In file included from /usr/include/c++/14.2.1/string:49,
                 from /usr/include/c++/14.2.1/stdexcept:39,
                 from /usr/include/boost/asio/io_context.hpp:20,
                 from /usr/include/boost/asio/io_service.hpp:18,
                 from /home/dormouse/gits/monero/src/p2p/net_node.h:34,
                 from /home/dormouse/gits/monero/src/rpc/instanciations.cpp:31:
/usr/include/c++/14.2.1/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
      |            ^~~~~~~~~~~~~~
In file included from /home/dormouse/gits/monero/contrib/epee/include/byte_slice.h:37,
                 from /home/dormouse/gits/monero/src/cryptonote_protocol/levin_notify.h:36,
                 from /home/dormouse/gits/monero/src/p2p/net_node.h:48:
/home/dormouse/gits/monero/contrib/epee/include/span.h: In instantiation of ‘epee::span<const unsigned char> epee::as_byte_span(const T&) [with T = boost::uuids::uuid]’:
/home/dormouse/gits/monero/contrib/epee/include/string_tools.h:94:39:   required from ‘std::string epee::string_tools::pod_to_hex(const t_pod_type&) [with t_pod_type = boost::uuids::uuid; std::string = std::__cxx11::basic_string<char>]’
   94 |     return to_hex::string(as_byte_span(s));
      |                           ~~~~~~~~~~~~^~~
/home/dormouse/gits/monero/src/cryptonote_protocol/cryptonote_protocol_handler.inl:320:57:   required from ‘std::__cxx11::list<cryptonote::connection_info> cryptonote::t_cryptonote_protocol_handler<t_core>::get_connections() [with t_core = cryptonote::core]’
  320 |       cnx.connection_id = epee::string_tools::pod_to_hex(cntxt.m_connection_id);
      |                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/rpc/instanciations.cpp:37:39:   required from here
   37 | namespace cryptonote { template class t_cryptonote_protocol_handler<cryptonote::core>; }
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/span.h:165:34: error: static assertion failed: source type may have padding
  165 |     static_assert(!has_padding<T>(), "source type may have padding");
      |                    ~~~~~~~~~~~~~~^~
/home/dormouse/gits/monero/contrib/epee/include/span.h:165:34: note: ‘! epee::has_padding<boost::uuids::uuid>()’ evaluates to false
In file included from /home/dormouse/gits/monero/contrib/epee/include/net/net_utils_base.h:40,
                 from /home/dormouse/gits/monero/contrib/epee/include/net/abstract_tcp_server2.h:58,
                 from /home/dormouse/gits/monero/src/p2p/net_node.h:50:
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h: In instantiation of ‘bool nodetool::basic_node_data::serialize_map(t_storage&, typename t_storage::hsection) [with bool is_store = true; t_storage = epee::serialization::portable_storage; typename t_storage::hsection = epee::serialization::section*]’:
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:194:5:   required from ‘bool nodetool::basic_node_data::store(t_storage&, typename t_storage::hsection) const [with t_storage = epee::serialization::portable_storage; typename t_storage::hsection = epee::serialization::section*]’
   50 |     return self.template serialize_map<true>(st, hparent_section); \
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization_overloads.h:73:23:   required from ‘bool epee::serialization::serialize_t_obj(const serializible_type&, t_storage&, typename t_storage::hsection, const char*) [with serializible_type = nodetool::basic_node_data; t_storage = portable_storage; typename t_storage::hsection = section*]’
   73 |       return obj.store(stg, hchild_section);
      |              ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization_overloads.h:260:31:   required from ‘static bool epee::serialization::kv_serialization_overloads_impl_is_base_serializable_types<false>::kv_serialize(const t_type&, t_storage&, typename t_storage::hsection, const char*) [with t_type = nodetool::basic_node_data; t_storage = epee::serialization::portable_storage; typename t_storage::hsection = epee::serialization::section*]’
  260 |         return serialize_t_obj(d, stg, hparent_section, pname);
      |                ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization_overloads.h:370:193:   required from ‘bool epee::serialization::kv_serialize(const t_type&, t_storage&, typename t_storage::hsection, const char*) [with t_type = nodetool::basic_node_data; t_storage = portable_storage; typename t_storage::hsection = section*]’
  370 |       return kv_serialization_overloads_impl_is_base_serializable_types<boost::mpl::contains<base_serializable_types<t_storage>, typename std::remove_const<t_type>::type>::value>::kv_serialize(d, stg, hparent_section, pname);
      |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization_overloads.h:329:28:   required from ‘static bool epee::serialization::selector<true>::serialize(const t_type&, t_storage&, typename t_storage::hsection, const char*) [with t_type = nodetool::basic_node_data; t_storage = epee::serialization::portable_storage; typename t_storage::hsection = epee::serialization::section*]’
  329 |         return kv_serialize(d, stg, hparent_section, pname);
      |                ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:221:9:   required from ‘bool nodetool::COMMAND_HANDSHAKE_T<t_playload_type>::request_t::serialize_map(t_storage&, typename t_storage::hsection) [with bool is_store = true; t_storage = epee::serialization::portable_storage; t_playload_type = cryptonote::CORE_SYNC_DATA; typename t_storage::hsection = epee::serialization::section*]’
   79 |   epee::serialization::selector<is_store>::serialize(this_ref.varialble, stg, hparent_section, val_name);
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:220:7:   required from ‘bool nodetool::COMMAND_HANDSHAKE_T<t_playload_type>::request_t::store(t_storage&, typename t_storage::hsection) const [with t_storage = epee::serialization::portable_storage; t_playload_type = cryptonote::CORE_SYNC_DATA; typename t_storage::hsection = epee::serialization::section*]’
   50 |     return self.template serialize_map<true>(st, hparent_section); \
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/storages/levin_abstract_invoke2.h:91:43:   required from ‘bool epee::net_utils::async_invoke_remote_command2(const connection_context_base&, int, const t_arg&, t_transport&, const callback_t&, size_t) [with t_result = epee::misc_utils::struct_init<nodetool::COMMAND_HANDSHAKE_T<cryptonote::CORE_SYNC_DATA>::response_t>; t_arg = epee::misc_utils::struct_init<nodetool::COMMAND_HANDSHAKE_T<cryptonote::CORE_SYNC_DATA>::request_t>; callback_t = nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core> >::do_handshake_with_peer(nodetool::peerid_type&, p2p_connection_context&, bool)::<lambda(int, const nodetool::COMMAND_HANDSHAKE_T<cryptonote::CORE_SYNC_DATA>::response&, nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core> >::p2p_connection_context&)>; t_transport = epee::levin::async_protocol_handler_config<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> >; size_t = long unsigned int]’
   91 |       const_cast<t_arg&>(out_struct).store(stg);//TODO: add true const support to searilzation
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
/home/dormouse/gits/monero/src/p2p/net_node.inl:1155:97:   required from ‘bool nodetool::node_server<t_payload_net_handler>::do_handshake_with_peer(nodetool::peerid_type&, p2p_connection_context&, bool) [with t_payload_net_handler = cryptonote::t_cryptonote_protocol_handler<cryptonote::core>; nodetool::peerid_type = long unsigned int; p2p_connection_context = nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context>]’
 1155 |     bool r = epee::net_utils::async_invoke_remote_command2<typename COMMAND_HANDSHAKE::response>(context_, COMMAND_HANDSHAKE::ID, arg, zone.m_net_server.get_config_object(),
      |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1156 |       [this, &pi, &ev, &hsh_result, &just_take_peerlist, &context_, &timeout](int code, const typename COMMAND_HANDSHAKE::response& rsp, p2p_connection_context& context)
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1157 |     {
      |     ~
 1158 |       epee::misc_utils::auto_scope_leave_caller scope_exit_handler = epee::misc_utils::create_scope_leave_handler([&](){ev.raise();});
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1159 |
      |
 1160 |       if(code < 0)
      |       ~~~~~~~~~~~~
 1161 |       {
      |       ~
 1162 |         LOG_WARNING_CC(context, "COMMAND_HANDSHAKE invoke failed. (" << code <<  ", " << epee::levin::get_err_descr(code) << ")");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1163 |         if (code == LEVIN_ERROR_CONNECTION_TIMEDOUT || code == LEVIN_ERROR_CONNECTION_DESTROYED)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1164 |           timeout = true;
      |           ~~~~~~~~~~~~~~~
 1165 |         return;
      |         ~~~~~~~
 1166 |       }
      |       ~
 1167 |
      |
 1168 |       if(rsp.node_data.network_id != m_network_id)
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1169 |       {
      |       ~
 1170 |         LOG_WARNING_CC(context, "COMMAND_HANDSHAKE Failed, wrong network!  (" << rsp.node_data.network_id << "), closing connection.");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1171 |         return;
      |         ~~~~~~~
 1172 |       }
      |       ~
 1173 |
      |
 1174 |       if(!handle_remote_peerlist(rsp.local_peerlist_new, context))
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1175 |       {
      |       ~
 1176 |         LOG_WARNING_CC(context, "COMMAND_HANDSHAKE: failed to handle_remote_peerlist(...), closing connection.");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1177 |         add_host_fail(context.m_remote_address);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1178 |         return;
      |         ~~~~~~~
 1179 |       }
      |       ~
 1180 |       hsh_result = true;
      |       ~~~~~~~~~~~~~~~~~~
 1181 |       if(!just_take_peerlist)
      |       ~~~~~~~~~~~~~~~~~~~~~~~
 1182 |       {
      |       ~
 1183 |         if(!m_payload_handler.process_payload_sync_data(rsp.payload_data, context, true))
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1184 |         {
      |         ~
 1185 |           LOG_WARNING_CC(context, "COMMAND_HANDSHAKE invoked, but process_payload_sync_data returned false, dropping connection.");
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1186 |           hsh_result = false;
      |           ~~~~~~~~~~~~~~~~~~~
 1187 |           return;
      |           ~~~~~~~
 1188 |         }
      |         ~
 1189 |
      |
 1190 |         pi = context.peer_id = rsp.node_data.peer_id;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1191 |         context.m_rpc_port = rsp.node_data.rpc_port;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1192 |         context.m_rpc_credits_per_hash = rsp.node_data.rpc_credits_per_hash;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1193 |         context.support_flags = rsp.node_data.support_flags;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1194 |         const auto azone = context.m_remote_address.get_zone();
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1195 |         network_zone& zone = m_network_zones.at(azone);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1196 |         zone.m_peerlist.set_peer_just_seen(rsp.node_data.peer_id, context.m_remote_address, context.m_pruning_seed, context.m_rpc_port, context.m_rpc_credits_per_hash);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1197 |
      |
 1198 |         // move
      |         ~~~~~~~
 1199 |         if(azone == epee::net_utils::zone::public_ && rsp.node_data.peer_id == zone.m_config.m_peer_id)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1200 |         {
      |         ~
 1201 |           LOG_DEBUG_CC(context, "Connection to self detected, dropping connection");
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1202 |           hsh_result = false;
      |           ~~~~~~~~~~~~~~~~~~~
 1203 |           return;
      |           ~~~~~~~
 1204 |         }
      |         ~
 1205 |         LOG_INFO_CC(context, "New connection handshaked, pruning seed " << epee::string_tools::to_string_hex(context.m_pruning_seed));
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1206 |         LOG_DEBUG_CC(context, " COMMAND_HANDSHAKE INVOKED OK");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1207 |       }else
      |       ~~~~~
 1208 |       {
      |       ~
 1209 |         LOG_DEBUG_CC(context, " COMMAND_HANDSHAKE(AND CLOSE) INVOKED OK");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1210 |       }
      |       ~
 1211 |       context_ = context;
      |       ~~~~~~~~~~~~~~~~~~~
 1212 |     }, P2P_DEFAULT_HANDSHAKE_INVOKE_TIMEOUT);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/rpc/instanciations.cpp:36:37:   required from here
   36 | namespace nodetool { template class node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core>>; }
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization.h:102:60: error: static assertion failed: t_type must be a POD type.
  102 |   static_assert(std::is_pod<decltype(this_ref.varialble)>::value, "t_type must be a POD type."); \
      |                                                            ^~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization.h:119:59: note: in expansion of macro ‘KV_SERIALIZE_VAL_POD_AS_BLOB_N’
  119 | #define KV_SERIALIZE_VAL_POD_AS_BLOB(varialble)           KV_SERIALIZE_VAL_POD_AS_BLOB_N(varialble, #varialble)
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:195:7: note: in expansion of macro ‘KV_SERIALIZE_VAL_POD_AS_BLOB’
  195 |       KV_SERIALIZE_VAL_POD_AS_BLOB(network_id)
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization.h:102:60: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
  102 |   static_assert(std::is_pod<decltype(this_ref.varialble)>::value, "t_type must be a POD type."); \
      |                                                            ^~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization.h:119:59: note: in expansion of macro ‘KV_SERIALIZE_VAL_POD_AS_BLOB_N’
  119 | #define KV_SERIALIZE_VAL_POD_AS_BLOB(varialble)           KV_SERIALIZE_VAL_POD_AS_BLOB_N(varialble, #varialble)
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:195:7: note: in expansion of macro ‘KV_SERIALIZE_VAL_POD_AS_BLOB’
  195 |       KV_SERIALIZE_VAL_POD_AS_BLOB(network_id)
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h: In instantiation of ‘bool nodetool::basic_node_data::serialize_map(t_storage&, typename t_storage::hsection) [with bool is_store = false; t_storage = epee::serialization::portable_storage; typename t_storage::hsection = epee::serialization::section*]’:
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:194:5:   required from ‘bool nodetool::basic_node_data::_load(t_storage&, typename t_storage::hsection) [with t_storage = epee::serialization::portable_storage; typename t_storage::hsection = epee::serialization::section*]’
   55 |     return serialize_map<false>(stg, hparent_section);\
      |            ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization_overloads.h:81:23:   required from ‘bool epee::serialization::unserialize_t_obj(serializible_type&, t_storage&, typename t_storage::hsection, const char*) [with serializible_type = nodetool::basic_node_data; t_storage = portable_storage; typename t_storage::hsection = section*]’
   81 |       return obj._load(stg, hchild_section);
      |              ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization_overloads.h:266:33:   required from ‘static bool epee::serialization::kv_serialization_overloads_impl_is_base_serializable_types<false>::kv_unserialize(t_type&, t_storage&, typename t_storage::hsection, const char*) [with t_type = nodetool::basic_node_data; t_storage = epee::serialization::portable_storage; typename t_storage::hsection = epee::serialization::section*]’
  266 |         return unserialize_t_obj(d, stg, hparent_section, pname);
      |                ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization_overloads.h:376:195:   required from ‘bool epee::serialization::kv_unserialize(t_type&, t_storage&, typename t_storage::hsection, const char*) [with t_type = nodetool::basic_node_data; t_storage = portable_storage; typename t_storage::hsection = section*]’
  376 |       return kv_serialization_overloads_impl_is_base_serializable_types<boost::mpl::contains<base_serializable_types<t_storage>, typename std::remove_const<t_type>::type>::value>::kv_unserialize(d, stg, hparent_section, pname);
      |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization_overloads.h:352:30:   required from ‘static bool epee::serialization::selector<false>::serialize(t_type&, t_storage&, typename t_storage::hsection, const char*) [with t_type = nodetool::basic_node_data; t_storage = epee::serialization::portable_storage; typename t_storage::hsection = epee::serialization::section*]’
  352 |         return kv_unserialize(d, stg, hparent_section, pname);
      |                ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:234:9:   required from ‘bool nodetool::COMMAND_HANDSHAKE_T<t_playload_type>::response_t::serialize_map(t_storage&, typename t_storage::hsection) [with bool is_store = false; t_storage = epee::serialization::portable_storage; t_playload_type = cryptonote::CORE_SYNC_DATA; typename t_storage::hsection = epee::serialization::section*]’
   79 |   epee::serialization::selector<is_store>::serialize(this_ref.varialble, stg, hparent_section, val_name);
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:233:7:   required from ‘bool nodetool::COMMAND_HANDSHAKE_T<t_playload_type>::response_t::load(t_storage&, typename t_storage::hsection) [with t_storage = epee::serialization::portable_storage; t_playload_type = cryptonote::CORE_SYNC_DATA; typename t_storage::hsection = epee::serialization::section*]’
   61 |     return serialize_map<false>(stg, hparent_section);\
      |            ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/storages/levin_abstract_invoke2.h:113:32:   required from ‘bool epee::net_utils::async_invoke_remote_command2(const connection_context_base&, int, const t_arg&, t_transport&, const callback_t&, size_t) [with t_result = epee::misc_utils::struct_init<nodetool::COMMAND_HANDSHAKE_T<cryptonote::CORE_SYNC_DATA>::response_t>; t_arg = epee::misc_utils::struct_init<nodetool::COMMAND_HANDSHAKE_T<cryptonote::CORE_SYNC_DATA>::request_t>; callback_t = nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core> >::do_handshake_with_peer(nodetool::peerid_type&, p2p_connection_context&, bool)::<lambda(int, const nodetool::COMMAND_HANDSHAKE_T<cryptonote::CORE_SYNC_DATA>::response&, nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core> >::p2p_connection_context&)>; t_transport = epee::levin::async_protocol_handler_config<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> >; size_t = long unsigned int]’
  113 |         if (!result_struct.load(stg_ret))
      |              ~~~~~~~~~~~~~~~~~~^~~~~~~~~
/home/dormouse/gits/monero/src/p2p/net_node.inl:1155:97:   required from ‘bool nodetool::node_server<t_payload_net_handler>::do_handshake_with_peer(nodetool::peerid_type&, p2p_connection_context&, bool) [with t_payload_net_handler = cryptonote::t_cryptonote_protocol_handler<cryptonote::core>; nodetool::peerid_type = long unsigned int; p2p_connection_context = nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context>]’
 1155 |     bool r = epee::net_utils::async_invoke_remote_command2<typename COMMAND_HANDSHAKE::response>(context_, COMMAND_HANDSHAKE::ID, arg, zone.m_net_server.get_config_object(),
      |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1156 |       [this, &pi, &ev, &hsh_result, &just_take_peerlist, &context_, &timeout](int code, const typename COMMAND_HANDSHAKE::response& rsp, p2p_connection_context& context)
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1157 |     {
      |     ~
 1158 |       epee::misc_utils::auto_scope_leave_caller scope_exit_handler = epee::misc_utils::create_scope_leave_handler([&](){ev.raise();});
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1159 |
      |
 1160 |       if(code < 0)
      |       ~~~~~~~~~~~~
 1161 |       {
      |       ~
 1162 |         LOG_WARNING_CC(context, "COMMAND_HANDSHAKE invoke failed. (" << code <<  ", " << epee::levin::get_err_descr(code) << ")");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1163 |         if (code == LEVIN_ERROR_CONNECTION_TIMEDOUT || code == LEVIN_ERROR_CONNECTION_DESTROYED)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1164 |           timeout = true;
      |           ~~~~~~~~~~~~~~~
 1165 |         return;
      |         ~~~~~~~
 1166 |       }
      |       ~
 1167 |
      |
 1168 |       if(rsp.node_data.network_id != m_network_id)
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1169 |       {
      |       ~
 1170 |         LOG_WARNING_CC(context, "COMMAND_HANDSHAKE Failed, wrong network!  (" << rsp.node_data.network_id << "), closing connection.");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1171 |         return;
      |         ~~~~~~~
 1172 |       }
      |       ~
 1173 |
      |
 1174 |       if(!handle_remote_peerlist(rsp.local_peerlist_new, context))
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1175 |       {
      |       ~
 1176 |         LOG_WARNING_CC(context, "COMMAND_HANDSHAKE: failed to handle_remote_peerlist(...), closing connection.");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1177 |         add_host_fail(context.m_remote_address);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1178 |         return;
      |         ~~~~~~~
 1179 |       }
      |       ~
 1180 |       hsh_result = true;
      |       ~~~~~~~~~~~~~~~~~~
 1181 |       if(!just_take_peerlist)
      |       ~~~~~~~~~~~~~~~~~~~~~~~
 1182 |       {
      |       ~
 1183 |         if(!m_payload_handler.process_payload_sync_data(rsp.payload_data, context, true))
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1184 |         {
      |         ~
 1185 |           LOG_WARNING_CC(context, "COMMAND_HANDSHAKE invoked, but process_payload_sync_data returned false, dropping connection.");
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1186 |           hsh_result = false;
      |           ~~~~~~~~~~~~~~~~~~~
 1187 |           return;
      |           ~~~~~~~
 1188 |         }
      |         ~
 1189 |
      |
 1190 |         pi = context.peer_id = rsp.node_data.peer_id;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1191 |         context.m_rpc_port = rsp.node_data.rpc_port;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1192 |         context.m_rpc_credits_per_hash = rsp.node_data.rpc_credits_per_hash;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1193 |         context.support_flags = rsp.node_data.support_flags;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1194 |         const auto azone = context.m_remote_address.get_zone();
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1195 |         network_zone& zone = m_network_zones.at(azone);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1196 |         zone.m_peerlist.set_peer_just_seen(rsp.node_data.peer_id, context.m_remote_address, context.m_pruning_seed, context.m_rpc_port, context.m_rpc_credits_per_hash);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1197 |
      |
 1198 |         // move
      |         ~~~~~~~
 1199 |         if(azone == epee::net_utils::zone::public_ && rsp.node_data.peer_id == zone.m_config.m_peer_id)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1200 |         {
      |         ~
 1201 |           LOG_DEBUG_CC(context, "Connection to self detected, dropping connection");
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1202 |           hsh_result = false;
      |           ~~~~~~~~~~~~~~~~~~~
 1203 |           return;
      |           ~~~~~~~
 1204 |         }
      |         ~
 1205 |         LOG_INFO_CC(context, "New connection handshaked, pruning seed " << epee::string_tools::to_string_hex(context.m_pruning_seed));
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1206 |         LOG_DEBUG_CC(context, " COMMAND_HANDSHAKE INVOKED OK");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1207 |       }else
      |       ~~~~~
 1208 |       {
      |       ~
 1209 |         LOG_DEBUG_CC(context, " COMMAND_HANDSHAKE(AND CLOSE) INVOKED OK");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1210 |       }
      |       ~
 1211 |       context_ = context;
      |       ~~~~~~~~~~~~~~~~~~~
 1212 |     }, P2P_DEFAULT_HANDSHAKE_INVOKE_TIMEOUT);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/rpc/instanciations.cpp:36:37:   required from here
   36 | namespace nodetool { template class node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core>>; }
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization.h:102:60: error: static assertion failed: t_type must be a POD type.
  102 |   static_assert(std::is_pod<decltype(this_ref.varialble)>::value, "t_type must be a POD type."); \
      |                                                            ^~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization.h:119:59: note: in expansion of macro ‘KV_SERIALIZE_VAL_POD_AS_BLOB_N’
  119 | #define KV_SERIALIZE_VAL_POD_AS_BLOB(varialble)           KV_SERIALIZE_VAL_POD_AS_BLOB_N(varialble, #varialble)
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:195:7: note: in expansion of macro ‘KV_SERIALIZE_VAL_POD_AS_BLOB’
  195 |       KV_SERIALIZE_VAL_POD_AS_BLOB(network_id)
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization.h:102:60: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
  102 |   static_assert(std::is_pod<decltype(this_ref.varialble)>::value, "t_type must be a POD type."); \
      |                                                            ^~~~~
/home/dormouse/gits/monero/contrib/epee/include/serialization/keyvalue_serialization.h:119:59: note: in expansion of macro ‘KV_SERIALIZE_VAL_POD_AS_BLOB_N’
  119 | #define KV_SERIALIZE_VAL_POD_AS_BLOB(varialble)           KV_SERIALIZE_VAL_POD_AS_BLOB_N(varialble, #varialble)
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/dormouse/gits/monero/src/p2p/p2p_protocol_defs.h:195:7: note: in expansion of macro ‘KV_SERIALIZE_VAL_POD_AS_BLOB’
  195 |       KV_SERIALIZE_VAL_POD_AS_BLOB(network_id)
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[3]: *** [src/rpc/CMakeFiles/obj_rpc.dir/build.make:149: src/rpc/CMakeFiles/obj_rpc.dir/instanciations.cpp.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:3562: src/rpc/CMakeFiles/obj_rpc.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
[ 52%] Building CXX object src/rpc/CMakeFiles/obj_daemon_rpc_server.dir/zmq_pub.cpp.o
In file included from /home/dormouse/gits/monero/src/cryptonote_basic/cryptonote_format_utils.h:33,
                 from /home/dormouse/gits/monero/src/rpc/zmq_pub.cpp:48:
/home/dormouse/gits/monero/src/cryptonote_basic/cryptonote_basic_impl.h:43:29: warning: ‘template<class _Arg, class _Result> struct std::unary_function’ is deprecated [-Wdeprecated-declarations]
   43 |   struct array_hasher: std::unary_function<t_array&, std::size_t>
      |                             ^~~~~~~~~~~~~~
In file included from /usr/include/c++/14.2.1/string:49,
                 from /usr/include/boost/assert/source_location.hpp:13,
                 from /usr/include/boost/exception/exception.hpp:9,
                 from /usr/include/boost/throw_exception.hpp:21,
                 from /usr/include/boost/thread/pthread/mutex.hpp:12,
                 from /usr/include/boost/thread/mutex.hpp:16,
                 from /home/dormouse/gits/monero/src/rpc/zmq_pub.h:33,
                 from /home/dormouse/gits/monero/src/rpc/zmq_pub.cpp:30:
/usr/include/c++/14.2.1/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
      |            ^~~~~~~~~~~~~~
[ 52%] Building CXX object src/rpc/CMakeFiles/obj_daemon_rpc_server.dir/zmq_server.cpp.o
make[2]: *** [CMakeFiles/Makefile2:4316: src/cryptonote_protocol/CMakeFiles/obj_cryptonote_protocol.dir/all] Error 2
[ 52%] Built target obj_daemon_messages
In file included from /home/dormouse/gits/monero/src/cryptonote_core/tx_pool.h:47,
                 from /home/dormouse/gits/monero/src/cryptonote_core/cryptonote_core.h:45,
                 from /home/dormouse/gits/monero/src/rpc/daemon_handler.h:35,
                 from /home/dormouse/gits/monero/src/rpc/daemon_handler.cpp:29:
/home/dormouse/gits/monero/src/cryptonote_basic/cryptonote_basic_impl.h:43:29: warning: ‘template<class _Arg, class _Result> struct std::unary_function’ is deprecated [-Wdeprecated-declarations]
   43 |   struct array_hasher: std::unary_function<t_array&, std::size_t>
      |                             ^~~~~~~~~~~~~~
In file included from /usr/include/c++/14.2.1/bits/unique_ptr.h:38,
                 from /usr/include/c++/14.2.1/memory:78,
                 from /home/dormouse/gits/monero/contrib/epee/include/byte_slice.h:33,
                 from /home/dormouse/gits/monero/src/rpc/daemon_handler.h:31:
/usr/include/c++/14.2.1/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
      |            ^~~~~~~~~~~~~~
[ 52%] Built target obj_daemon_rpc_server
make[1]: *** [Makefile:146: all] Error 2
make[1]: Leaving directory '/home/dormouse/gits/monero/build/Linux/_HEAD_detached_at_v0.18.3.4_/release'
make: *** [Makefile:103: release-all] Error 2
```

I have disabled threads for this one for better readability.

```
-- Configuring done (10.1s)
-- Generating done (1.0s)
-- Build files have been written to: /home/dormouse/gits/monero/build/Linux/_HEAD_detached_at_v0.18.3.4_/release
make[1]: Entering directory '/home/dormouse/gits/monero/build/Linux/_HEAD_detached_at_v0.18.3.4_/release'
[  2%] Built target generate_translations_header
[  5%] Built target libminiupnpc-static
[  5%] Building C object external/miniupnp/miniupnpc/CMakeFiles/upnpc-static.dir/upnpc.c.o
[  5%] Linking C executable upnpc-static
[  5%] Built target upnpc-static
[  7%] Built target libminiupnpc-shared
[  8%] Building C object external/miniupnp/miniupnpc/CMakeFiles/upnpc-shared.dir/upnpc.c.o
[  8%] Linking C executable upnpc-shared
[  8%] Built target upnpc-shared
[  8%] Building C object external/miniupnp/miniupnpc/CMakeFiles/listdevices.dir/listdevices.c.o
/home/dormouse/gits/monero/external/miniupnp/miniupnpc/listdevices.c: In function ‘add_device’:
/home/dormouse/gits/monero/external/miniupnp/miniupnpc/listdevices.c:60:24: error: implicit declaration of function ‘strdup’; did you mean ‘strcmp’? [-Wimplicit-function-declaration]
   60 |         elt->descURL = strdup(dev->descURL);
      |                        ^~~~~~
      |                        strcmp
/home/dormouse/gits/monero/external/miniupnp/miniupnpc/listdevices.c:60:22: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   60 |         elt->descURL = strdup(dev->descURL);
      |                      ^
make[3]: *** [external/miniupnp/miniupnpc/CMakeFiles/listdevices.dir/build.make:79: external/miniupnp/miniupnpc/CMakeFiles/listdevices.dir/listdevices.c.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:1713: external/miniupnp/miniupnpc/CMakeFiles/listdevices.dir/all] Error 2
make[1]: *** [Makefile:146: all] Error 2
make[1]: Leaving directory '/home/dormouse/gits/monero/build/Linux/_HEAD_detached_at_v0.18.3.4_/release'
make: *** [Makefile:103: release-all] Error 2
```

@tankf33der This looks like an issue in the source tree to me.
```
/home/dormouse/gits/monero/external/miniupnp/miniupnpc/listdevices.c:60:24: error: implicit declaration of function ‘strdup’; did you mean ‘strcmp’? [-Wimplicit-function-declaration]
   60 |         elt->descURL = strdup(dev->descURL);
      |                        ^~~~~~
      |                        strcmp
/home/dormouse/gits/monero/external/miniupnp/miniupnpc/listdevices.c:60:22: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   60 |         elt->descURL = strdup(dev->descURL);
      |                      ^
```

--- a/external/CMakeLists.txt
+++ b/external/CMakeLists.txt
@@ -39,6 +39,7 @@ find_package(Miniupnpc REQUIRED)
 
 message(STATUS "Using in-tree miniupnpc")
 set(UPNPC_NO_INSTALL TRUE CACHE BOOL "Disable miniupnp installation" FORCE)
+set(UPNPC_BUILD_SHARED OFF CACHE BOOL "Disable building shared library" FORCE)
 add_subdirectory(miniupnp/miniupnpc)
 set_property(TARGET libminiupnpc-static PROPERTY FOLDER "external")
 set_property(TARGET libminiupnpc-static PROPERTY POSITION_INDEPENDENT_CODE ON)

